SampleProviders/WFATSampleProvider.jl

```diff
@@ -0,0 +1,165 @@
+    ss_pzSamples    ::AbstractVector;
+    ionRatePrefix   ::Symbol;       # currently supports :ExpRate.
+    tunExit         ::Symbol;       # :Para for tunneling, :IpF for over-barrier, automatically specified.
+    mol_Ip          ::Real;
+    mol_μ           ::Vector;
+    wfat_intdata    ::Array;
+    wfat_nξMax      ::Int;
+    wfat_mMax       ::Int;
+    wfat_lMax       ::Int;
+
+    function WFATSampleProvider(;
+                                laser               ::Laser,
+                                target              ::Molecule,
+                                sample_tSpan        ::Tuple{<:Real,<:Real},
+                                sample_tSampleNum   ::Int,
+                                rate_ionRatePrefix  ::Symbol,
+                                ss_pdMax            ::Real,
+                                ss_pdNum            ::Int,
+                                ss_pzMax            ::Real,
+                                ss_pzNum            ::Int,
+                                mol_ionOrbitRelHOMO ::Int,
+                                kwargs...   # kwargs are surplus params.
+                                )
+        # check sampling parameters.
+        @assert (sample_tSampleNum>0) "[WFATSampleProvider] Invalid time sample number $sample_tSampleNum."
+        @assert (ss_pdNum>0 && ss_pzNum>0) "[WFATSampleProvider] Invalid pd/pz sample number $ss_pdNum/$ss_pzNum."
+        # load WFAT IntData.
+        if ! (mol_ionOrbitRelHOMO in MolWFATAvailableIndices(target))
+            MolCalcWFATData!(target, mol_ionOrbitRelHOMO)
+        end
+        μ, intdata = MolWFATData(target, mol_ionOrbitRelHOMO)
+        Ip = IonPotential(target, mol_ionOrbitRelHOMO)
+        nξMax = size(intdata,1) - 1
+        mMax = round(Int,(size(intdata,2)-1)/2)
+        lMax = size(intdata,3) - 1
+        # check IonRate prefix support.
+        if ! (rate_ionRatePrefix in [:ExpRate])
+            error("[WFATSampleProvider] Undefined tunneling rate prefix [$rate_ionRatePrefix].")
+        end
+        # check Keldysh parameter & over-barrier condition.
+        F0 = LaserF0(laser)
+        γ0 = AngFreq(laser) * sqrt(2Ip) / F0
+        if γ0 ≥ 0.5
+            @warn "[WFATSampleProvider] Keldysh parameter γ=$γ0, adiabatic (tunneling) condition [γ<<1] not sufficiently satisfied."
+        elseif γ0 ≥ 1.0
+            @warn "[WFATSampleProvider] Keldysh parameter γ=$γ0, adiabatic (tunneling) condition [γ<<1] unsatisfied."
+        end
+        F_crit = Ip^2/4/(1-sqrt(Ip/2))
+        tunExit = :Para
+        if F0 ≥ F_crit*2/3
+            @warn "[WFATSampleProvider] Peak electric field strength F0=$F0, reaching 2/3 of over-barrier critical value, weak-field condition not sufficiently satisfied."
+        elseif F0 ≥ F_crit
+            @warn "[WFATSampleProvider] Peak electric field strength F0=$F0, reaching the over-barrier critical value, weak-field condition unsatisfied."
+            tunExit = :IpF
```

Comment:
Warn user that you have changed tunExit set by them.

Reply:
Finished in [db8191e].